---
title: "BMENE4000 Assignment 1 (Dataset 2)"
author: "Michael Yang, my2699"
Due date: "10/14/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

## Installation
```{r}
if(!requireNamespace("BiocManager",quietly=TRUE))
  install.packages("BiocManager")
BiocManager::install(version="3.13")
```


```{r}
BiocManager::install(c('BiocGenerics','DelayedArray','DelayedMatrixStats','limma','S4Vectors','SingleCellExperiment','SummarizedExperiment','batchelor','Matrix.utils'))
```


```{r}
library(dplyr)
library(ggplot2)
library(monocle3)
```

```{r}
cds <- readRDS("dataset2.rds")
```

View containers in the cell_data_set object:

Using pipes:
```{r}
cds %>% exprs() %>% head()
```

Dataframes:
```{r}
cds %>% colData() %>% head()
```
Columns:
sample, cell, total_umis, size_factor, gene, sgRNA, sgRNA_umi_count, proportion, condition, treatment, position 


```{r}
cds %>% rowData() %>% head()
```
Columns: ID, gene_short_name


Analyze the types of data:
```{r}
cds %>% exprs() %>% class()
cds %>% colData() %>% class()
cds %>% rowData() %>% class()
```


Summarize using dplyr
```{r}
cds %>%
  colData() %>%
  as.data.frame() %>%
  group_by(sample) %>%
  summarise(total_cells = n())
```

```{r}
rows <- cds %>%
  rowData() %>%
  as.data.frame() %>%
  group_by(gene_short_name) %>%
  summarise(total_genes = n())
rows
```

## Q1-1:
```{r}
plot_df <- cds %>% colData() %>% as.data.frame()
p <- ggplot(plot_df, aes(x = sample, y = total_umis, fill = sample), fun = median())+
  geom_violin() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
p + stat_summary(fun=median, geom="point") + ggtitle('Unique Transcripts Per Cell')
```

```{r}
cds %>%
  colData() %>%
  as.data.frame() %>%
  group_by(sample) %>%
  summarise(Median_UMIs = median(total_umis))
```

There are 4 unique samples in the data set (aggregated samples), as after using dplyr to "group by" samples, the final summary data frame shows only 4 rows.
I found the 'median number' of UMIs (I used the column 'total_umis') by passing through the 'median()' function in summarise, as well as plotted it on the violin plots. The median UMIs per cell for each sample is shown in the table above (12576, 13264, 14576, 16550.5 for samples 1, 2, 3, 4 respectively). 



## Q1-2:

Creating a new cds table with the columns added by detect_genes function in monocle3, 'num_genes_expressed' and 'num_cells_expressed' as mentioned in the documentation.
```{r}
g <- detect_genes(cds, min_expr = 0)
g %>% colData() %>% head()
```

Median number of genes quantified per cell for each sample:
```{r}
g %>%
  colData() %>%
  as.data.frame() %>%
  group_by(sample) %>%
  summarise(Median = median(num_genes_expressed))
```


```{r}
plot_df <- g %>% colData() %>% as.data.frame()
p <- ggplot(plot_df, aes(x = sample, y = num_genes_expressed, fill = sample), fun = median())+
  geom_violin() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
p + stat_summary(fun=median, geom="point")
```
```{r}
sum(g$num_genes_expressed)
```
```{r}
g %>%
  rowData() %>%
  as.data.frame() %>%
  group_by(gene_short_name) %>%
  summarise(Num_genes = n())
```

The number of genes expressed can be taken by summing the column 'num_genes_expressed' in the detect_genes-altered cds, as that seems to collect the total number of expressed genes in the dataset. That number should be approximately 21151828 genes total. Based on the summary data on the rowData() by grouping by unique IDs, we have 32,6438 rows of unique gene short names, meaning we have approximately that many unique genes. I used gene_short_name as the count in summarise as it is possible for multiople ENSEMBL gene IDs to map to one gene name. 
The median values for number of genes per cell for each sample are displayed in the table above generated by the function summarise(median(num_genes_expressed)), 2961, 3216, 3333, 3667.5 for samples 1, 2, 3, 4 respectively.


## Q1-3: Summary of Unique Treatments
```{r}
cds %>%
  colData() %>%
  as.data.frame() %>%
  group_by(treatment) %>%
  summarise(total_cells = n())
```

Based on the above code for a summary of unique treatments in the data summarized by total cell counts per treatment, we can see that there are 2 unique treatments present in the dataset, "mock" and "tgfb", and there are 2867 and 3695 cells for each respective treatment. These translate to mock-control group and TGFB, which likely is TGF-beta.


### Part 2: Dataset Exploration

```{r}
library(ggpubr)
```

## Q2-1: Genes as Function of Treatment
```{r}
cds_subset <- cds[rowData(cds)$gene_short_name == 'MKI67',]
plot_genes_violin(cds_subset, group_cells_by = 'treatment') + stat_compare_means(aes(label = ..p.signif..), method = 't.test')
```

```{r}
cds_subset <- cds[rowData(cds)$gene_short_name == 'CDK1',]
plot_genes_violin(cds_subset, group_cells_by = 'treatment') + stat_compare_means(aes(label = ..p.signif..), method = 't.test')
```
From using the function stat_compare_means to test for significance between the mean of gene expression as function of treatment, we can see that for the 'MKI67', there is no significance, but for the 'CDK1' gene, the three asterisks *** signify a p value < 0.001, which indicates significance in the difference between the mean expressions between mock treatment and tgfb. Therefore, while the plots don't seem to show a very apparent effect of the treatment on gene expression, there is statistical significance of the difference for the CDK1 gene (and not the MKI67 gene). 

Therefore, we can conclude that the tgfb treatment does not have a significant effect on the expression of MKI67, which in turn does not affect proliferation; however, we can see that as the differences in treatment were significant, and the expressions of CDK1 genes were lower in the 'tgfb' treatment group compared to 'mock' treatment, indicating that tgfb treatment can likely decrease the prevalence of the mitosis phenotype. Interestingly, it has been reported that in Yamamoto et al, 2020, interactions between CDK1 and the TGFB pathway have not been completely explored as of yet. 
Source: "Cyclin Dependent Kinase 1 (CDK1) positively regulates cardiac hypertrophy and fibrosis via TGF-beta pathway", European Heart Journal, Yamamoto et al., 2020


## Q2-2: Mechanism of Action for TGFB
The treatment that is not mock or control in the dataset, 'tgfb', likely refers to TGFB, which is the transforming growth factor beta (TGF-beta) which can control aspects of cellular behavior such as proliferation. More specifically, in almost all non-neoplastic epithelial cells, TGFB stimulation will induce cytostasis (inhibition of cell growth and multiplication), as well as in endothelial cells, hematopoietic cells, neuronal cells and certain mesenchymal cells. It also promotes proliferation in certain mesenchymal cells. Additionally, it can either induce or suppress apoptosis in different cell types. Most of the information used to answer this question was referenced from:
"TGF-B family signalling in the control of cell proliferation and survival", Zhang, Y., Alexander, P.B., & Wang, X.F., Cold Spring Harbor Perspectives in Biology, 2017, 9(4)
TGFB signaling is initiated when an activated TGFB-ligand interacts with TGFB type 2 receptor and type 1 receptors to form a complex, which then results in activation of type 1 receptors and allows binding of receptor regulated R-Smads. The activated R-Smads then move into the nucleus and form transcriptional complexes with Smad4 and other DNA-binding cofactors, non-DNA binding coactivators, and corepressors, which then will cooperatively regulate expression of target genes. Depending on several factors such as the number of original TGFB ligands activated, the intracellular expression pattern of receptor/Smad interacting proteins, and other factors, the signaling pathway can lead to a large number of gene responses (regulation of expressions). In terms of cancer cells, TGFB can be an immunosuppressor and pro-angiogenic factor; over-expression of TGFB-1 has been linked to various cancers in the forms of tumor progression and metastasis; therefore, blocking the TGFB pathway have been essential to fighting cancer.



## Q2-3: Plotting of Expression of CDH1 and FN1
```{r}
cds_subset <- cds[rowData(cds)$gene_short_name == 'CDH1',]
plot_genes_violin(cds_subset, group_cells_by = 'treatment') + stat_compare_means(aes(label = ..p.signif..), method = 't.test')
```

```{r}
cds_subset <- cds[rowData(cds)$gene_short_name == 'FN1',]
plot_genes_violin(cds_subset, group_cells_by = 'treatment') + stat_compare_means(aes(label = ..p.signif..), method = 't.test')
```
From plotting the expression levels with these two marker genes associated with response to tgfb treatment, we can see that CDH1 expression actually did not show significance, while the differences in FN1 gene expression did have statistically significant differences, with p < 0.00001 (****). Within literature, it has been reported that FN1 (fibronectin 1) has indeed increased in certain cases of lymphatic endothelial cells when TGF-B (in that case, TGF-B2) was used as treatment on the cells (Yoshimatsu et al., 2020)
Source: "TGF-Beta and TNF-alpha cooperatively induce mesenchymal transition of lymphatic endothelial cells via activation of Activin signals", Yoshimatsu et al., PLOS One, 1-28, 2020

In terms of whether the change in FN1 expression is affected by TGFB treatment is consistent, it seems to make sense considering how TGFB treatment affects cell proliferation. FN1 is the sole gene that encodes for fibronectins, which are glycoproteins in the ECM that are involved in key cellular processes such as cell growth, differentiation, adhesion, or migration (Ventura et al., 2018), which is definitely a pathway that TGFB can affect as mentioned previously. It has been shown that TGFB signaling has indeed resulted in a positive correlation with expression of FNs in tumor vasculature.
Source: "TGF-B induces oncofetal fibronectin that, in turn, modulates TGF-B superfamily signaling in endothelial cells", Ventura et al., Journal of Cell Science, 2018, 131(1)

CDH1 (cadherin 1) is a gene that is critical in epithelial cell adherence, and changes in expression/ mutations / loss of function of this gene have been linked to cancer progression by increasing proliferation, invasion, and metastasis. It has been shown to be linked to the TGFB pathway, but in this case here there is no significant changes in CDH1 expression under TGFB treatment compared to mock group. Therefore, it seems strange that in this case there is no association between the mechanism of treatment and the expression of CDH1. 


## Q2-4: Plotting of Expression of New Gene
```{r}
cds_subset <- cds[rowData(cds)$gene_short_name == 'TGM2',]
plot_genes_violin(cds_subset, group_cells_by = 'treatment') + stat_compare_means(aes(label = ..p.signif..), method = 't.test')
```
An additional gene pathway that is associated with TGFB treatment that had an observed result in cells is TGM2, or transglutaminase 2, which is a protein coding gene that primarily mediates post-translational modifications of proteins into protein-bound glutamine residues. It also been observed to increase in expression in several types of cancers, especially so in cancer-associated stress conditions (Cho et al., 2020). In a study analyzing the expression profiling of genes regulated by TGF-beta in cancer cells, TGM2 was notably increased in expression for a long period of time following TGFB treatment, even compared to expression of other genes such as MMP2 or IGFBP7 (Ranganathan et al., 2007). 

We can see here that the difference in gene expression of TGM2 between the mock control and the TGFB treatment groups is very significant, with TGM2 expression increasing greatly in the non-control treatment group, with statistical significance and p-value < 0.0001 (noted by the four asterisks, ****, when analyzing the mean expressions between the 2 groups). This upregulation in TGM2 gene expression following TGFB treatment is in-line with literature that also observes this.

Sources:
"Amplification of transglutaminase 2 enhances tumor-promoting inflammation in gastric cancers", Cho et al., Experimental and Molecular Medicine, 2020, 52, 854  - 864;
"Expression profiling of genes regulated by TGF-beta: Differential regulation in normal and tumour cells", Ranganathan et al., BMC Genomics, 2007, 8(98)


### Question 3: Concepts from Class

## Q3-1: NGS
As the genomic region has approximately 10kb in length, it would be best to use Single Molecule Sequencing, which can read long reads up to 15 kb with medium throughput, and can process hundreds of fragments at the same time. 

In terms of sequencing methods that we have learned in class, I think that nanopore sequencing would be a good choice to increase the probability of obtaining a correct genome assembly for the region. The problem of repetitive sequences in the strand make sequencing particularly challenging, as short reads and assembly from that would not be able to properly characterize the sequence, especially if there are SNPs or other variations as short reads can lead to confusion of mapping the variation to the sequence when they are in sections (Treangen & Salzberg, 2012). Meanwhile, nanopore sequencing is a long-read method, and able to read entire sequences multiple times in quick succession, and in real-time, which can help mitigate the problems that short read techniques can bring. This method works by running a DNA sequence through a nanopore channel with a motor protein, and an unwinding enzyme (helicase) will separate the DNA, and depending on the base that goes through the channel, the ionic current that is being run through the nanopore exit will be disrupted. This will give a real-time sequencing of the bases that pass through by measuring the change in ionic current through time. As we discussed in class, this technique can be done in quick repetition with high-throughput, which can increase accuracy and thus confidence in the sequence that we get (and confirm the correct sequence including the repetitive sequences in our genomic region). 


## Q3-2: GWAS
From class, we learned that the Bonferroni correction method in multiple hypothesis testing to test for true SNPs with significant associations refers to the equation, "0.05 divided by the number of measurements" to find the threshold p-value for significance. In this case, our number of measurements is 1,000,000 SNPs (1 x 10^6). Therefore, based on the Bonferroni correction method, we should be looking for significance only if the p-value for the SNP locus is below (0.05/1,000,000) = 5 x 10^-8, or 0.00000005. From the two given p-values, we can see that BHMT, with the p-value of 5.8 x 10^-24, is the only one of the two which can then be considered a significant hit. 

As noted from class, genome-wide association studies are mainly used to compare genetic variants in large numbers of affected cases to those in unaffected controls to determine whether an association between a gene and a phenotype such as atherosclerosis exists. Based on this, we cannot say that GWAS studies alone are completely confirmatory that a certain SNP or gene is involved in the disease, but mainly that there is an association with the presence/distribution of the gene/SNP and the disease/trait based on studies between the diseased population and the controls (by identifying them as genomic markers). These associated markers are linked to QTLs, or quantitative trait loci, of that trait. From the result, though, we can treat these areas as areas of interest, and we can hypothesize that the SNP is in linkage disequilibrium with a QTL that is associated with the disease. We can use Direct QTL Mapping or Association Mapping to identify loci associated with the trait. We can also validate SNPs of interest first by using technology to isolate them, such as Affymetrix SNP arrays or Illumina's SNP microarrays, then using statistical tests such as Chi square tests of independence to test "how much are observations deviating from our expectations, given the the SNP vs. the original allele, total number of each type of SNP in each population and total number of cases and controls"? This will give us a more quantitative probability that the change we observe is due to chance. By doing these statistical tests for multiple SNPs in the dataset of interest, then establishing a p-value threshold to determine significance of association (corrected with Bonferroni, etc.), then we can establish SNPs or genes that are involved in the disease. 
